File: kotlin/examples/kotlin/Group.kt
// Copyright AGNTCY Contributors (https://github.com/agntcy)
// SPDX-License-Identifier: Apache-2.0

package io.agntcy.slim.examples

import io.agntcy.slim.examples.common.*
import io.agntcy.slim.bindings.*
import kotlinx.coroutines.*
import java.time.Duration
import java.util.concurrent.atomic.AtomicReference

/**
 * Group example (heavily commented).
 * 
 * Purpose:
 *   Demonstrates how to:
 *     * Start a local Slim app using the global service
 *     * Optionally create a group session (becoming its moderator)
 *     * Invite other participants (by their IDs) into the group
 *     * Receive and display messages
 *     * Interactively publish messages
 * 
 * Key concepts:
 *   - Group sessions are created with SessionConfig with SessionType.GROUP and
 *     reference a 'topic' (channel) Name.
 *   - Invites are explicit: the moderator invites each participant after
 *     creating the session.
 *   - Participants that did not create the session simply wait for
 *     listenForSessionAsync() to yield their Session.
 * 
 * Usage:
 *   Group --local org/default/me --remote org/default/chat-topic \
 *         --invites org/default/peer1 --invites org/default/peer2
 * 
 * Notes:
 *   * If --invites is omitted, the client runs in passive participant mode.
 *   * If both remote and invites are supplied, the client acts as session moderator.
 */

/**
 * Handle inviting a participant to the group.
 */
suspend fun handleInvite(session: Session, inviteId: String) {
    val parts = inviteId.trim().split(" ")
    if (parts.size != 1) {
        println("Error: 'invite' command expects exactly one participant ID (e.g., 'invite org/ns/client-1')")
        return
    }
    
    println("Inviting participant: $inviteId")
    val inviteName = Name.fromString(inviteId)
    try {
        val handle = session.inviteAsync(inviteName)
        handle.waitAsync()
        println("✅ Successfully invited $inviteId")
    } catch (e: Exception) {
        val errorStr = e.message ?: ""
        when {
            errorStr.contains("participant already in group", ignoreCase = true) -> {
                println("Error: Participant $inviteId is already in the group.")
            }
            errorStr.contains("failed to add participant to session", ignoreCase = true) -> {
                println("Error: Failed to add participant $inviteId to session.")
            }
            else -> throw e
        }
    }
}

/**
 * Handle removing a participant from the group.
 */
suspend fun handleRemove(session: Session, removeId: String) {
    val parts = removeId.trim().split(" ")
    if (parts.size != 1) {
        println("Error: 'remove' command expects exactly one participant ID (e.g., 'remove org/ns/client-1')")
        return
    }
    
    println("Removing participant: $removeId")
    val removeName = Name.fromString(removeId)
    try {
        val handle = session.removeAsync(removeName)
        handle.waitAsync()
        println("✅ Successfully removed $removeId")
    } catch (e: Exception) {
        val errorStr = e.message ?: ""
        when {
            errorStr.contains("participant not found in group", ignoreCase = true) -> {
                println("Error: Participant $removeId is not in the group.")
            }
            else -> throw e
        }
    }
}

/**
 * Receive messages for the bound session.
 * 
 * Behavior:
 *   * If not moderator: wait for a new group session (listenForSessionAsync()).
 *   * If moderator: reuse the createdSession reference.
 *   * Loop forever until cancellation or an error occurs.
 */
suspend fun receiveLoop(
    localApp: App,
    createdSession: Session?,
    sessionReady: CompletableDeferred<Session>
) {
    val session = createdSession ?: run {
        println("Waiting for session...")
        localApp.listenForSessionAsync(null)
    }
    
    // Signal that session is ready
    sessionReady.complete(session)
    
    // Get source name for display
    val sourceName = session.source()
    
    while (true) {
        try {
            // Await next inbound message from the group session
            val receivedMsg = session.getMessageAsync(Duration.ofSeconds(30))
            val ctx = receivedMsg.context
            val payload = receivedMsg.payload
            
            // Display sender name and message (using the actual sender from the context)
            val sender = ctx.sourceName.toIdString()
            println("$sender > ${String(payload)}")
            
            // if the message metadata contains PUBLISH_TO this message is a reply
            // to a previous one. In this case we do not reply to avoid loops
            if (!ctx.metadata.containsKey("PUBLISH_TO")) {
                val reply = "message received by ${sourceName.toIdString()}"
                session.publishToAsync(ctx, reply.toByteArray(), null, ctx.metadata)
            }
        } catch (e: CancellationException) {
            // Graceful shutdown path (ctrl-c or program exit)
            break
        } catch (e: Exception) {
            // Break if session is closed, otherwise continue listening
            if (e.message?.contains("session closed", ignoreCase = true) == true) {
                break
            }
            continue
        }
    }
}

/**
 * Interactive loop allowing participants to publish messages.
 * 
 * Typing 'exit' or 'quit' (case-insensitive) terminates the loop.
 * Typing 'remove NAME' removes a participant from the group
 * Typing 'invite NAME' invites a participant to the group
 * Each line is published to the group channel as UTF-8 bytes.
 */
suspend fun keyboardLoop(
    createdSession: Session?,
    sessionReady: CompletableDeferred<Session>,
    localApp: App
) {
    try {
        // Wait for the session to be established
        val session = sessionReady.await()
        val destName = session.destination()
        
        if (createdSession != null) {
            println("Welcome to the group ${destName.toIdString()}!")
            println("Commands:")
            println("  - Type a message to send it to the group")
            println("  - 'remove NAME' to remove a participant")
            println("  - 'invite NAME' to invite a participant")
            println("  - 'exit' or 'quit' to leave the group")
        } else {
            println("Welcome to the group ${destName.toIdString()}!")
            println("Commands:")
            println("  - Type a message to send it to the group")
            println("  - 'exit' or 'quit' to leave the group")
        }
        
        while (true) {
            val userInput = readlnOrNull() ?: break
            
            when {
                userInput.trim().lowercase() in listOf("exit", "quit") -> {
                    if (createdSession != null) {
                        // Delete the session
                        val handle = localApp.deleteSessionAsync(session)
                        handle.waitAsync()
                    }
                    break
                }
                
                userInput.trim().lowercase().startsWith("invite ") && createdSession != null -> {
                    val inviteId = userInput.substring(7).trim()
                    handleInvite(session, inviteId)
                    continue
                }
                
                userInput.trim().lowercase().startsWith("remove ") && createdSession != null -> {
                    val removeId = userInput.substring(7).trim()
                    handleRemove(session, removeId)
                    continue
                }
                
                else -> {
                    // Send message to the channel_name specified when creating the session.
                    // As the session is group, all participants will receive it.
                    session.publishAsync(userInput.toByteArray(), null, null)
                }
            }
        }
    } catch (e: CancellationException) {
        // Handle task cancellation gracefully
    } catch (e: Exception) {
        println("-> Error sending message: ${e.message}")
    }
}

/**
 * Orchestrate one group-capable client instance.
 * 
 * Modes:
 *   * Moderator (creator): remote (channel) + invites provided.
 *   * Listener only: no remote; waits for inbound group sessions.
 * 
 * @param config GroupConfig instance containing all configuration
 */
suspend fun runClient(config: GroupConfig) = coroutineScope {
    // Create the local Slim instance using global service
    val (localApp, connId) = createLocalApp(config)
    
    // Parse the remote channel/topic if provided; else None triggers passive mode
    val chatChannel = config.remote?.let { Name.fromString(it) }
    
    // Session sharing between tasks
    val sessionReady = CompletableDeferred<Session>()
    
    // Session object only exists immediately if we are moderator
    var createdSession: Session? = null
    if (chatChannel != null && config.invites != null) {
        // We are the moderator; create the group session now
        Colors.printFormatted(
            "Creating new group session (moderator)... ${Name.fromString(config.local).toIdString()}"
        )
        
        // Create group session configuration
        val sessionConfig = SessionConfig(
            sessionType = SessionType.GROUP,
            enableMls = config.enableMls,
            maxRetries = 5u,
            interval = Duration.ofSeconds(5),
            metadata = emptyMap()
        )
        
        // Create session - returns a context with completion and session
        val sessionContext = localApp.createSession(sessionConfig, chatChannel)
        // Wait for session to be established
        sessionContext.completion.waitAsync()
        createdSession = sessionContext.session
        
        // Invite each provided participant
        config.invites.forEach { invite ->
            val inviteName = Name.fromString(invite)
            localApp.setRouteAsync(inviteName, connId)
            val handle = createdSession.inviteAsync(inviteName)
            handle.waitAsync()
            println("${config.local} -> add ${inviteName.toIdString()} to the group")
        }
    }
    
    // Launch the receiver and keyboard loop on separate threads (parallel execution)
    val receiveJob = launch(Dispatchers.IO) {
        receiveLoop(localApp, createdSession, sessionReady)
    }
    
    val keyboardJob = launch(Dispatchers.IO) {
        keyboardLoop(createdSession, sessionReady, localApp)
    }
    
    // Wait for keyboard loop to finish (user typed exit)
    keyboardJob.join()
    
    // Cancel the receive loop
    receiveJob.cancel()
}

/**
 * Main entry point for the group example.
 * 
 * Parses command-line arguments and runs the client.
 */
fun main(args: Array<String>) = runBlocking {
    try {
        val config = ConfigParser.parseGroupArgs(args)
        
        // Run the client
        runClient(config)
    } catch (e: IllegalArgumentException) {
        println("Configuration error: ${e.message}")
        println()
        println("Usage: Group --local <org/ns/app> [OPTIONS]")
        println()
        println("Options:")
        println("  --local <id>              Local ID (org/namespace/app) - REQUIRED")
        println("  --remote <id>             Remote ID (org/namespace/channel)")
        println("  --slim <url>              SLIM server endpoint (default: http://127.0.0.1:46357)")
        println("  --invites <id>            Invite participant (can be specified multiple times)")
        println("  --shared-secret <secret>  Shared secret for authentication")
        println("  --enable-mls              Enable MLS encryption")
        println("  --enable-opentelemetry    Enable OpenTelemetry tracing")
        kotlin.system.exitProcess(1)
    } catch (e: CancellationException) {
        println("\nClient interrupted by user.")
    } catch (e: Exception) {
        println("Error: ${e.message}")
        e.printStackTrace()
        kotlin.system.exitProcess(1)
    }
}

File: node/examples/group.ts
// Copyright AGNTCY Contributors (https://github.com/agntcy)
// SPDX-License-Identifier: Apache-2.0

/**
 * SLIM Group Messaging Example
 *
 * Demonstrates multicast (group) sessions:
 *   - Moderator mode: create a group session for a channel and invite participants.
 *   - Participant mode: wait for an incoming group session invitation.
 *   - Both modes then exchange messages interactively over the group session.
 *
 * Based on implementations from the Go, Python, and .NET bindings.
 */

// @ts-expect-error - tsx resolves .js imports to .ts files at runtime
import * as slimBindings from '../generated/index.js';
import * as readline from 'node:readline/promises';
import { createAndConnectApp, splitId, logMessage, sleep, DEFAULT_SERVER_ENDPOINT, DEFAULT_SHARED_SECRET } from './common.js';

const DEFAULT_LOCAL_ID = 'org/default/me';

/**
 * Command line arguments
 */
interface CliArgs {
  local: string;
  remote?: string;
  server: string;
  sharedSecret: string;
  enableMls: boolean;
  invites: string[];
}

/**
 * Parse command line arguments
 */
function parseArgs(): CliArgs {
  const args = process.argv.slice(2);
  let local = DEFAULT_LOCAL_ID;
  let remote: string | undefined;
  let server = DEFAULT_SERVER_ENDPOINT;
  let sharedSecret = DEFAULT_SHARED_SECRET;
  let enableMls = false;
  let invites: string[] = [];

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--local' && i + 1 < args.length) {
      local = args[i + 1];
      i++;
    } else if (args[i] === '--remote' && i + 1 < args.length) {
      remote = args[i + 1];
      i++;
    } else if (args[i] === '--server' && i + 1 < args.length) {
      server = args[i + 1];
      i++;
    } else if (args[i] === '--secret' && i + 1 < args.length) {
      sharedSecret = args[i + 1];
      i++;
    } else if (args[i] === '--enable-mls') {
      enableMls = true;
    } else if (args[i] === '--invites' && i + 1 < args.length) {
      invites = args[i + 1]
        .split(',')
        .map((s) => s.trim())
        .filter((s) => s.length > 0);
      i++;
    } else if (args[i] === '--help' || args[i] === '-h') {
      console.log('Usage: npm run group -- [OPTIONS]');
      console.log('');
      console.log('Options:');
      console.log('  --local <id>          Local identity (default: org/default/me)');
      console.log('  --remote <id>         Remote channel/topic identity (moderator mode)');
      console.log('  --server <address>    Server endpoint (default: http://localhost:46357)');
      console.log('  --secret <secret>     Shared secret (default: demo-shared-secret-min-32-chars!!)');
      console.log('  --enable-mls          Enable MLS encryption');
      console.log('  --invites <ids>       Comma-separated participant IDs to invite (moderator mode)');
      console.log('  --help, -h            Show this help message');
      console.log('');
      console.log('If --remote and --invites are both supplied, this instance creates and moderates');
      console.log('the group session. Otherwise it waits to be invited into one.');
      process.exit(0);
    }
  }

  return { local, remote, server, sharedSecret, enableMls, invites };
}

/** Extract a readable message from a thrown SlimError/UniffiError-shaped value. */
function describeError(error: any): string {
  if (error?.tag && error?.inner?.message) {
    return error.inner.message;
  }
  return error?.message || String(error);
}

/**
 * Run as moderator: create a group session for the channel and invite participants.
 */
async function runModerator(
  app: any,
  connId: bigint,
  remote: string,
  invites: string[],
  enableMls: boolean,
  instance: string
): Promise<void> {
  const channelName = splitId(remote);

  logMessage(instance, `📡 Creating group session as moderator for channel: ${remote}`);

  const config = {
    sessionType: slimBindings.SessionType.Group,
    enableMls,
    maxRetries: 5,
    interval: 5000,
    metadata: new Map(),
  };

  const session = await app.createSessionAndWaitAsync(config, channelName);

  // Give session a moment to establish
  await sleep(100);
  logMessage(instance, '✅ Group session created');

  for (const inviteId of invites) {
    try {
      const inviteName = splitId(inviteId);
      app.setRoute(inviteName, connId);
      await session.inviteAndWaitAsync(inviteName);
      logMessage(instance, `👥 Invited ${inviteId} to the group`);
    } catch (error: any) {
      logMessage(instance, `⚠️  Failed to invite ${inviteId}: ${describeError(error)}`);
    }
  }

  try {
    await runMessageLoops(session, channelName, instance);
  } finally {
    try {
      const handle = await app.deleteSessionAsync(session);
      await handle.waitAsync();
    } catch {
      // Session may already be closed - nothing more to clean up.
    }
  }
}

/**
 * Run as participant: wait for an incoming group session invitation.
 */
async function runParticipant(app: any, instance: string): Promise<void> {
  logMessage(instance, '👂 Waiting for incoming group session invitation...');

  const timeout = 60000;
  const session = await app.listenForSessionAsync(timeout);
  const channelName = session.destination();

  logMessage(instance, `🎉 Joined group session for channel: ${channelName}`);

  try {
    await runMessageLoops(session, channelName, instance);
  } finally {
    try {
      const handle = await app.deleteSessionAsync(session);
      await handle.waitAsync();
    } catch {
      // Session may already be closed - nothing more to clean up.
    }
  }
}

/**
 * Run the receive loop and the interactive keyboard loop concurrently until
 * either one finishes (session ends, or the user exits).
 */
async function runMessageLoops(session: any, channelName: any, instance: string): Promise<void> {
  const controller = new AbortController();
  const receivePromise = receiveLoop(session, instance, controller);
  const keyboardPromise = keyboardLoop(session, channelName, instance, controller);

  try {
    await Promise.race([receivePromise, keyboardPromise]);
  } finally {
    controller.abort();
    await Promise.allSettled([receivePromise, keyboardPromise]);
  }
}

/**
 * Receive messages for the group session and auto-reply (unless the message
 * is itself already a reply, indicated by a PUBLISH_TO metadata entry, to
 * avoid an infinite reply loop).
 */
async function receiveLoop(session: any, instance: string, controller: AbortController): Promise<void> {
  const sourceName = session.source();

  while (!controller.signal.aborted) {
    let received: any;
    try {
      received = await session.getMessageAsync(1000, { signal: controller.signal });
    } catch (error: any) {
      if (controller.signal.aborted) {
        return;
      }
      const msg = describeError(error);
      if (msg.toLowerCase().includes('timeout')) {
        continue;
      }
      logMessage(instance, `🔚 Session ended: ${msg}`);
      controller.abort();
      return;
    }

    const ctx = received.context;
    const text = Buffer.from(received.payload).toString('utf-8');
    console.log(`\n${ctx.sourceName} > ${text}`);

    // If this message already carries PUBLISH_TO, it is itself a reply - don't reply again.
    if (!ctx.metadata.has('PUBLISH_TO')) {
      const reply = `message received by ${sourceName}`;
      try {
        await session.publishToAndWaitAsync(ctx, Buffer.from(reply), undefined, ctx.metadata);
      } catch (error: any) {
        logMessage(instance, `⚠️  Error sending auto-reply: ${describeError(error)}`);
      }
    }

    process.stdout.write(`${sourceName} > `);
  }
}

/**
 * Interactive loop: type a message and press Enter to send it to the group;
 * 'exit' or 'quit' leaves the group.
 */
async function keyboardLoop(session: any, channelName: any, instance: string, controller: AbortController): Promise<void> {
  const sourceName = session.source();
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

  try {
    console.log(`\nWelcome to the group ${channelName}!`);

    const participants: any[] = await session.participantsListAsync();
    logMessage(instance, `👥 Participants in the group (${participants.length}):`);
    for (const participant of participants) {
      logMessage(instance, `   👤 ${participant}`);
    }

    console.log(`Type a message and press Enter to send, or 'exit'/'quit' to leave.\n`);

    while (!controller.signal.aborted) {
      let input: string;
      try {
        input = await rl.question(`${sourceName} > `, { signal: controller.signal });
      } catch {
        return;
      }

      const trimmed = input.trim();
      if (trimmed.length === 0) {
        continue;
      }
      if (trimmed.toLowerCase() === 'exit' || trimmed.toLowerCase() === 'quit') {
        logMessage(instance, '👋 Leaving group...');
        return;
      }

      try {
        await session.publishAndWaitAsync(Buffer.from(trimmed), undefined, undefined);
        logMessage(instance, '📤 Sent to group');
      } catch (error: any) {
        logMessage(instance, `❌ Error sending message: ${describeError(error)}`);
      }
    }
  } finally {
    rl.close();
  }
}

/**
 * Main entry point
 */
async function main() {
  const args = parseArgs();

  console.log('👥 SLIM Group Messaging Example');
  console.log('================================');
  console.log(`Local ID: ${args.local}`);
  if (args.remote) console.log(`Remote (channel): ${args.remote}`);
  console.log(`Server: ${args.server}`);
  if (args.invites.length > 0) console.log(`Invites: ${args.invites.join(', ')}`);
  console.log();

  try {
    const { app, connId } = await createAndConnectApp(args.local, args.server, args.sharedSecret);
    const instance = app.id();

    const isModerator = Boolean(args.remote) && args.invites.length > 0;
    if (isModerator) {
      await runModerator(app, connId, args.remote!, args.invites, args.enableMls, instance);
    } else {
      await runParticipant(app, instance);
    }
  } catch (error: any) {
    console.error('Fatal error:', describeError(error));
    process.exit(1);
  }
}

// Handle signals gracefully
process.on('SIGINT', () => {
  console.log('\n\n📋 Received SIGINT');
  console.log('🛑 Shutting down...');
  process.exit(0);
});

process.on('SIGTERM', () => {
  console.log('\n\n📋 Received SIGTERM');
  console.log('🛑 Shutting down...');
  process.exit(0);
});

// Run the main function
main();
